public/app/core/components/RolePicker/RolePickerMenu.tsx

```diff
@@ -0,0 +1,585 @@
+import React, { FormEvent, useCallback, useEffect, useState } from 'react';
+import { css, cx } from '@emotion/css';
+import {
+  Button,
+  Checkbox,
+  CustomScrollbar,
+  HorizontalGroup,
+  RadioButtonGroup,
+  Tooltip,
+  useStyles2,
+  useTheme2,
+} from '@grafana/ui';
+import { GrafanaTheme2, SelectableValue } from '@grafana/data';
+import { getSelectStyles } from '@grafana/ui/src/components/Select/getSelectStyles';
+import { OrgRole, Role } from 'app/types';
+
+type BuiltInRoles = { [key: string]: Role[] };
+
+const BuiltinRoles = ['Viewer', 'Editor', 'Admin'];
+const BuiltinRoleOption: Array<SelectableValue<OrgRole>> = BuiltinRoles.map((r: string) => ({
+  label: r,
+  value: r as OrgRole,
+}));
+
+const fixedRoleGroupNames: { [key: string]: string } = {
+  ldap: 'LDAP',
+  current: 'Current org',
+};
+
+interface RolePickerMenuProps {
+  builtInRole: OrgRole;
+  builtInRoles: BuiltInRoles;
+  options: Role[];
+  appliedRoles: Role[];
+  showGroups?: boolean;
+  onSelect: (roles: Role[]) => void;
+  onBuiltInRoleSelect?: (role: OrgRole) => void;
+  onUpdate: (newBuiltInRole: OrgRole, newRoles: string[]) => void;
+  onClear?: () => void;
+}
+
+export const RolePickerMenu = ({
+  builtInRole,
+  builtInRoles,
+  options,
+  appliedRoles,
+  showGroups,
+  onSelect,
+  onBuiltInRoleSelect,
+  onUpdate,
+  onClear,
+}: RolePickerMenuProps): JSX.Element => {
+  const [selectedOptions, setSelectedOptions] = useState<Role[]>(appliedRoles);
+  const [selectedBuiltInRole, setSelectedBuiltInRole] = useState<OrgRole>(builtInRole);
+  const [selectedBuiltInRoles, setSelectedBuiltInRoles] = useState<Role[]>(builtInRoles[builtInRole]);
+  const [showSubMenu, setShowSubMenu] = useState(false);
+  const [openedMenuGroup, setOpenedMenuGroup] = useState('');
+  const [subMenuOptions, setSubMenuOptions] = useState<Role[]>([]);
+
+  const theme = useTheme2();
+  const styles = getSelectStyles(theme);
+  const customStyles = useStyles2(getStyles);
+
+  const getOptionGroups = useCallback(() => {
```

Comment:
This could be a utility function defined outside of the component. This would remove the need for `useCallback`.